### PR TITLE
Bump `mapbox-navigation-native` dependency version to `51.0.0`

### PR DESCRIPTION
**Pull request by Guardiola31337**
### Description

Bumps `mapbox-navigation-native` dependency version to `51.0.0` and fixes breaking changes

### Changelog

```
<changelog>Removed `getRouteGeometryWithBuffer` API from `RouteProgress`.</changelog>
<changelog>:warning: `MapboxNavigation` history recording APIs `retrieveHistory`, `toggleHistory` and `addHistoryEvent` are no-op in this release.</changelog>
```

cc @mskurydin @etl @vanmaxim

### CODE REVIEW

File: examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryRecorder.kt [hunk 1]
```diff
@@ -43,7 +43,6 @@ class HistoryRecorder(
      */
     fun startRecording() {
         startedAt = Date()
-        navigation.toggleHistory(true)
     }
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/core/replay/HistoryRecorder.kt:47]: This needs a whole overhaul as History recording APIs changed drastically cc @kmadsen

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 2]
```diff
@@ -1065,6 +1064,7 @@ class MapboxNavigation(
                 USER_AGENT,
                 BuildConfig.NAV_NATIVE_SDK_VERSION,
                 isFallback,
+                "",
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1067]: @mskurydin @korshaknn Could you check this when you have a chance? 🙏

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 3]
```diff
@@ -168,10 +168,10 @@ internal class MapboxTripSession(
                 }
             }
 
-            override fun onCanReturnToLatest() {
+            override fun onCanReturnToLatest(version: String) {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:171]: @korshaknn @mskurydin could you check this when you have a chance? 🙏

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 4]
```diff
@@ -916,7 +914,7 @@ class MapboxNavigation(
             }
         }
 
-        override fun onCanReturnToLatest() {
+        override fun onCanReturnToLatest(version: String) {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:917]: Same as above. @korshaknn @mskurydin could you check this when you have a chance? 🙏

**Comment by Guardiola31337** [comment]
History recording APIs and tooling need a whole overhaul as NN `51.0.0` introduced new `HistoryRecorderHandle` and `HistoryReader` built on top of PBF.

cc @kmadsen @LukasPaczos @mskurydin @vanmaxim @etl @truburt @zugaldia

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 4 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:917]: this one can be renamed to `onCanReturnToTarget` (rename on NN side)

**Comment by Guardiola31337** [comment]
For reference I used the following `diff` to test history recording 👇 

```diff
diff --git a/examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt b/examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt
index acc7ab8f3..4290415a5 100644
--- a/examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt
+++ b/examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt
@@ -5,6 +5,7 @@ import android.content.res.Configuration
 import android.content.res.Resources
 import android.location.Location
 import android.os.Bundle
+import android.util.Log
 import android.view.View.INVISIBLE
 import android.view.View.VISIBLE
 import androidx.appcompat.app.AppCompatActivity
@@ -70,6 +71,7 @@ import com.mapbox.navigation.ui.voice.model.SpeechVolume
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
+import java.io.File
 import java.util.Locale
 
 class MapboxNavigationActivity : AppCompatActivity() {
@@ -300,10 +302,12 @@ class MapboxNavigationActivity : AppCompatActivity() {
             enabled = true
         }
 
+        val historyDir = File(filesDir, "history")
         // initialize Mapbox Navigation
         mapboxNavigation = MapboxNavigation(
             NavigationOptions.Builder(this)
                 .accessToken(getMapboxAccessTokenFromResources())
+                .historyDir(historyDir.path)
                 .build()
         )
         // move the camera to current location on the first update
@@ -447,6 +451,9 @@ class MapboxNavigationActivity : AppCompatActivity() {
                 binding.soundButton.unmuteAndExtend(2000L)
                 voiceInstructionsPlayer.volume(SpeechVolume(1f))
             }
+            mapboxNavigation.retrieveHistory {
+                Log.e("DEBUG", "history file is available in $it")
+            }
         }
 
         // start the trip session to being receiving location updates in free drive
```

Testing this noticed that an extra empty history file is generated after closing a session. This is probably because how `HistoryRecorderHandle` works internally and the fact that recording is always active:

```java
/**
 * HistoryRecorder works in auto-recording mode. The behaviour is the following:
 * History starts recording right after `HistoryRecorder` creation. Calling `dumpHistory` you could stop the current
 * recording session and get all files available for uploading through `HistoryDumpCallback`.
 * Recording will automatically starts again.
 */
 ```
 
 @mskurydin @vanmaxim Is it expected to have an empty file if you don't finally call `dumpHistory`? Could you check that internally and remove empty files when the session is over?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 2 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1067]:
```suggestion
                navigationOptions.routingTilesOptions.tilesVersion,
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 4 above)
Comment (korshaknn) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:917]: сс @mskurydin

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 5]
```diff
@@ -1065,6 +1061,7 @@ class MapboxNavigation(
                 USER_AGENT,
                 BuildConfig.NAV_NATIVE_SDK_VERSION,
                 isFallback,
+                navigationOptions.routingTilesOptions.tilesVersion,
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1064]: Are we passing the version twice? How's this different than the `tilesVersion` param?

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1064]: cc @mskurydin @korshaknn

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1064]: I think they could be different if fallback is used https://github.com/mapbox/mapbox-navigation-android/blob/f0399b6d1fd04adc25c0434f6ad8eec7a34ad938/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L899-L905

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1064]: Reading the `TileEndpointConfiguration` we seem to have `version` and `versionBeforeFallback` :+1:

nit: maybe both should be passed as params for clarity?

Comment (mskurydin) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1064]: We may specify one version in the endpoint configuration that corresponds to a chosen off-line region and the original version here to return to (usually, the original version).

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1064]: BTW @korshaknn suggested that above https://github.com/mapbox/mapbox-navigation-android/pull/4478#discussion_r645620565

**Review by LukasPaczos — APPROVED** [review]
LGTM

**Comment by vanmaxim** [comment]
>Is it expected to have an empty file if you don't finally call dumpHistory? Could you check that internally and remove empty files when the session is over?
I'll make an issue and fix it soon.

**Pull request merged**